src/ImageSharp/Formats/Tiff/TiffBitsPerSampleExtensions.cs

```diff
@@ -13,27 +13,42 @@ internal static class TiffBitsPerSampleExtensions
         /// </summary>
         /// <param name="tiffBitsPerSample">The tiff bits per sample.</param>
         /// <returns>Bits per sample array.</returns>
-        public static ushort[] Bits(this TiffBitsPerSample tiffBitsPerSample)
+        public static ushort[] BitsPerChannel(this TiffBitsPerSample tiffBitsPerSample)
```

Comment:
You're looking to avoid having to do this and also support any arbitrary arrays of up to 3 elements with each value maxed at 16 right?

Couldn't we define a struct made up of 3 `ushort` fields? and have `TryParse(ushort[])` and `ToArray()` properties. The array allocations would be so small an uncommon they wouldn't be an issue.

Something like this?

```c#
public readonly struct TiffBitsPerSample : IEquatable<TiffBitsPerSample>
{
    public readonly ushort Channel0;
    public readonly ushort Channel1;
    public readonly ushort Channel2;

    public TiffBitsPerSample(ushort channel0, ushort channel1, ushort channel2)
    {
        this.Channel0 = (ushort)Numerics.Clamp(channel0, 1, 16);
        this.Channel1 = (ushort)Numerics.Clamp(channel1, 0, 16);
        this.Channel2 = (ushort)Numerics.Clamp(channel2, 0, 16);
    }

    public static bool TryParse(ushort[] value, out TiffBitsPerSample sample)
    {
        if (value is null || value.Length == 0)
        {
            sample = default;
            return false;
        }

        ushort c2;
        ushort c1;
        ushort c0;
        switch (value.Length)
        {
            case 3:
                c2 = value[2];
                c1 = value[1];
                c0 = value[0];
                break;
            case 2:
                c2 = 0;
                c1 = value[1];
                c0 = value[0];
                break;
            default:
                c2 = 0;
                c1 = 0;
                c0 = value[0];
                break;
        }

        sample = new TiffBitsPerSample(c0, c1, c2);
        return true;
    }

    public override bool Equals(object obj)
        => obj is TiffBitsPerSample sample && this.Equals(sample);

    public bool Equals(TiffBitsPerSample other)
        => this.Channel0 == other.Channel0
        && this.Channel1 == other.Channel1
        && this.Channel2 == other.Channel2;

    public override int GetHashCode()
        => HashCode.Combine(this.Channel0, this.Channel1, this.Channel2);

    public ushort[] ToArray()
    {
        if (this.Channel1 == 0)
        {
            return new ushort[] { this.Channel0 };
        }

        if (this.Channel2 == 0)
        {
            return new ushort[] { this.Channel0, this.Channel1 };
        }

        return new ushort[] { this.Channel0, this.Channel1, this.Channel2 };
    }

    public override readonly string ToString()
        => $"TiffBitsPerSample({this.Channel0}, {this.Channel1}, {this.Channel2})";
}
```

You could then create static instances of known types for easy comparison during encoding.

Reply:
Good idea, I have changed BitsPerSample now to a struct aa848d74.

> You're looking to avoid having to do this and also support any arbitrary arrays of up to 3 elements with each value maxed at 16 right?

Arbitrary is maybe strong word here. As Ildar said, the spec allows arbitrary values, but i have not yet seen a tiff in the wild which has not the same values for all channels. The maximum is 32 bits per channel, so we currently only missing 24 and 32 bits per channel as far as i can tell.

Reply:
Glad I could help here!